text/0000-guaranteed-slice-repr.md

```diff
@@ -0,0 +1,205 @@
+- Feature Name: guaranteed_slice_repr
+- Start Date: 2025-02-18
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC guarantees the in-memory representation of slice and str references.
+Specifically, `&[T]` and `&mut [T]` are guaranteed to have the same layout as:
+
+```rust
+#[repr(C)]
+struct Slice<T> {
+    data: *const T,
+    len: usize,
```

Comment:
I think it's better to use `NonNull` here:
```rust
struct Slice<T> {
    data: NonNull<T>,
    len: usize,
}
```
`NonNull` is FFI-safe and provides the niche, so `Option<&[T]>` and `Option<Slice<T>>` would behave the same way.

Reply:
"same layout as" here means field order, alignment and size, not validity. Validity requirements are mentioned further down. `len` also isn't really `usize` since it has invalid values which we may exploit in the future.